File: priv/combined_metrics/samples/code_smells/no_fixme_comments/bad/importer.js
// Handles importing data from CSV and external sources.

const fs = require("fs");

// FIXME: this crashes on empty files, need to handle that
function importCsv(path) {
  const content = fs.readFileSync(path, "utf8");
  return content
    .split("\n")
    .map(parseRow)
    .filter((row) => row !== null);
}

// TODO: FIXME - validate headers before parsing rows
function parseRow(line) {
  const parts = line.split(",");
  if (parts.length === 3) {
    const [id, name, email] = parts;
    return { id, name, email };
  }
  // XXX: silently drops malformed rows, should log or collect errors
  return null;
}

function importUsers(rows) {
  // FIXME: this does N+1 inserts, wrap in a transaction
  return rows.map((row) => insertUser(row));
}

function validateRow(row) {
  // XXX: email regex is wrong, doesn't handle subdomains
  if (row.email.includes("@")) {
    return { ok: row };
  }
  return { error: "invalid_email" };
}

function deduplicate(rows) {
  // FIXME: uses email as dedup key but doesn't normalize case first
  const seen = new Map();
  for (const row of rows) {
    if (!seen.has(row.email)) {
      seen.set(row.email, row);
    }
  }
  return Array.from(seen.values());
}

function importFromApi(sourceUrl) {
  // TODO: FIXME - add retry logic and timeout handling
  const fetched = fetchRemote(sourceUrl);
  if (fetched) {
    return parseApiResponse(fetched);
  }
  // XXX: swallows all errors, need proper error propagation
  return [];
}

function transformRow(row, fieldMap) {
  // FIXME: doesn't handle nested fields or type coercion
  const result = {};
  for (const [src, dst] of Object.entries(fieldMap)) {
    result[dst] = row[src];
  }
  return result;
}

function writeResults(results, outputPath) {
  // XXX: overwrites file without backup, could lose data
  const content = results.map(formatResult).join("\n");
  fs.writeFileSync(outputPath, content);
}

function insertUser(row) {
  return { ok: row };
}

function fetchRemote(_url) {
  return [];
}

function parseApiResponse(data) {
  return data;
}

function formatResult(result) {
  return JSON.stringify(result);
}

module.exports = {
  importCsv,
  parseRow,
  importUsers,
  validateRow,
  deduplicate,
  importFromApi,
  transformRow,
  writeResults,
};

File: priv/combined_metrics/samples/code_smells/no_fixme_comments/bad/importer.py
"""Handles importing data from CSV and external sources."""

from pathlib import Path


# FIXME: this crashes on empty files, need to handle that
def import_csv(path):
    text = Path(path).read_text()
    rows = []
    for line in text.split("\n"):
        parsed = parse_row(line)
        if parsed is not None:
            rows.append(parsed)
    return rows


# TODO: FIXME - validate headers before parsing rows
def parse_row(line):
    parts = line.split(",")
    if len(parts) == 3:
        row_id, name, email = parts
        return {"id": row_id, "name": name, "email": email}
    # XXX: silently drops malformed rows, should log or collect errors
    return None


def import_users(rows):
    # FIXME: this does N+1 inserts, wrap in a transaction
    return [insert_user(row) for row in rows]


def validate_row(row):
    # XXX: email regex is wrong, doesn't handle subdomains
    if "@" in row["email"]:
        return {"ok": row}
    return {"error": "invalid_email"}


def deduplicate(rows):
    # FIXME: uses email as dedup key but doesn't normalize case first
    seen = {}
    for row in rows:
        if row["email"] not in seen:
            seen[row["email"]] = row
    return list(seen.values())


def import_from_api(source_url):
    # TODO: FIXME - add retry logic and timeout handling
    fetched = fetch_remote(source_url)
    if fetched is not None:
        return parse_api_response(fetched)
    # XXX: swallows all errors, need proper error propagation
    return []


def transform_row(row, field_map):
    # FIXME: doesn't handle nested fields or type coercion
    return {dst: row.get(src) for src, dst in field_map.items()}


def write_results(results, output_path):
    # XXX: overwrites file without backup, could lose data
    content = "\n".join(format_result(r) for r in results)
    Path(output_path).write_text(content)


def insert_user(row):
    return {"ok": row}


def fetch_remote(_url):
    return []


def parse_api_response(data):
    return data


def format_result(result):
    return repr(result)
